File: pset01/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Forward Declarations
int ValidateCardLong(int digit_long);
int *Long2Array(long card_no, int digit_long);
int LuhnSum(int *card_no, int c);

// Check if the digits long of the card is valid
int ValidateCardLong(int digit_long)
{
    if (digit_long < 13 || digit_long > 16 || digit_long == 14)
    {
        printf("INVALID\n");
        exit(0);
    }
    return 1;
}

// Convert format of the card number from a long to an array
int *Long2Array(long card_no, int digit_long)
{
    int static card_array[18];
    while (card_no != 0)
    {
        card_array[digit_long - 1] = card_no % 10;
        card_no /= 10;
        digit_long--;
    }
    return card_array;
}

// An algorithm invented by Hans Peter Luhn of IBM to checksum the credit card number
int LuhnSum(int *card_no, int digit_long)
{
    int sum = 0;
    for (int i = digit_long - 2; i >= 0; i -= 2)
    {
        if (card_no[i] * 2 >= 10)
        {
            int x = card_no[i] * 2;
            while (x != 0)
            {
                sum += x % 10;
                x /= 10;
            }
        }
        else
        {
            sum += card_no[i] * 2;
        }
    }
    for (int i = digit_long - 1; i >= 0; i -= 2)
    {
        sum += card_no[i];
    }
    return sum;
}

int main(void)
{
    long card_no = get_long("Number: ");
    int digit_long = floor(log10(card_no)) + 1;

    ValidateCardLong(digit_long);
    int *card_array = Long2Array(card_no, digit_long);

    int sum = LuhnSum(card_array, digit_long);
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        exit(0);
    }

    if (
        digit_long == 15 &&
        card_array[0] == 3 &&
        (card_array[1] == 4 || card_array[1] == 7))
    {
        printf("AMEX\n");
    }
    else if ((digit_long == 13 || digit_long == 16) && card_array[0] == 4)
    {
        printf("VISA\n");
    }
    else if (
        digit_long == 16 &&
        card_array[0] == 5 &&
        (card_array[1] == 1 ||
         card_array[1] == 2 ||
         card_array[1] == 3 ||
         card_array[1] == 4 ||
         card_array[1] == 5))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset06/sentimental-credit/credit.py
import sys


def main():
    """main()"""

    while True:
        str_card = input("Number: ")
        try:
            card = [int(i) for i in str_card]
        except ValueError:
            continue
        if card:
            break

    if len(card) < 13 or len(card) > 16 or check_luh_sum(card) is False:
        print("INVALID")
        sys.exit(0)

    if len(card) == 15 and card[0] == 3 and card[1] in (4, 7):
        print("AMEX")
    elif len(card) in (13, 16) and card[0] == 4:
        print("VISA")
    elif len(card) == 16 and card[0] == 5 and card[1] in range(1, 6):
        print("MASTERCARD")
    else:
        print("INVALID")


def check_luh_sum(card) -> bool:
    """
    An algorithm invented by Hans Peter Luhn of IBM to checksum the 
    credit card number
    """
    lsum = 0
    for i in range(len(card) - 2, -1, -2):
        if card[i] * 2 > 9:
            lsum = lsum + 1 + card[i] * 2 - 10
            continue
        lsum += card[i] * 2

    for i in range(len(card) -1 , -1, -2):
        lsum += card[i]

    if lsum % 10 == 0:
        return True
    else:
        return False


if __name__ == "__main__":
    main()
